### PR TITLE
EVG-18682: Upgrade to Ubuntu 20.04

### PR DESCRIPTION
**Pull request by sophstad**
EVG-18682

### Description

- Upgrade build variant to Ubuntu 20.04
- Install [NVM](https://github.com/nvm-sh/nvm) in `/data/<id>/.nvm` for each task. The installation is fully local to the task directory and won't affect any other tasks that run on these hosts. Sadly this adds about 1 minute to each task, so the overall makespan for our testing is longer. [BUILD-16733](https://jira.mongodb.org/browse/BUILD-16733) is open with the ask to have NVM preinstalled on Ubuntu.
  - I opted to specify our Node version in our `.evergreen.yml`. I think there's a benefit to having it tracked via git and managed there rather than as a variable.
- Remove `attach-server-logs` as this never matched any files on recent runs. It seemed old and not relevant anymore.
- Increase heap size when running Storybook as I started seeing OOM errors again. Hopefully upgrading Vite and Storybook will fix this (EVG-18849, EVG-18874).
- Combine `cli-scripts/` and `scripts/` directories... I couldn't identify a meaningful difference between them but let me know if there is 🤷‍♀️
- I'll remove unused variables in project settings after these changes are applied.
- Updated the git tag version definition in project settings so that our deploy script should run on any version of Ubuntu:
  image

### Testing
- Still looking into the failing e2e test but feel free to review!

### CODE REVIEW

**Review by minnakt — COMMENTED** [review]
I think one unfortunate side effect of moving those functions out of `post` is that test results / files won't get uploaded if the task fails. For example here's a [patch](https://spruce.mongodb.com/version/63dc772e7742ae5282e3436a/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) where I ran a few different tasks and you can see that the tests / files corresponding to each task didn't get attached. I guess the task just fails and exits before getting to the upload step?

Depending on what we want to do this could be okay, but just wanted to point it out

File: .evergreen.yml [hunk 1]
```diff
@@ -50,6 +54,29 @@ functions:
...
+    - command: shell.exec
+      params:
+        script: |
+          ${PREPARE_SHELL}
+
+          git clone https://github.com/nvm-sh/nvm.git "$NVM_DIR"
+          cd "$NVM_DIR"
+          git checkout `git describe --abbrev=0 --tags --match "v[0-9]*" $(git rev-list --tags --max-count=1)`
+          [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+          cd -
```
Comment (khelif96) [inline .evergreen.yml:75]: Can you explain what this does?

File: .evergreen.yml [hunk 2]
```diff
@@ -50,6 +54,29 @@ functions:
...
 
+  setup-node:
+    - command: subprocess.exec
+      params:
+        binary: bash
+        args:
+          - spruce/scripts/prepare-shell.sh
+    - command: expansions.update
+      params:
+        file: expansion.yml
```
Comment (khelif96) [inline .evergreen.yml:65]: Great job using all of the evergreen features!

File: .evergreen.yml [hunk 3]
```diff
@@ -167,54 +193,41 @@ functions:
...
-      display_name: "video:"
-
-  attach-server-logs:
-    command: s3.put
-    type: system
-    params:
-      aws_key: ${aws_key}
-      aws_secret: ${aws_secret}
-      local_files_include_filter:
-        ["evergreen/server_logs.txt"]
```
Comment (khelif96) [inline .evergreen.yml:212]: I think there may have been a point in time when the evergreen log output was piped to a file. Logkeeper does this rn.

File: cypress/integration/hosts/hosts_sorting.ts [hunk 4]
```diff
@@ -15,7 +15,8 @@ const sortByTests = [
       "build10.ny.cbi.10gen.cc",
     ],
   },
-  {
+  /* TODO: Re-enable distro sort upon completion of EVG-18900
```
Comment (khelif96) [inline cypress/integration/hosts/hosts_sorting.ts:18]: Can we ticket this and link it to EVG-18900 please.

File: .evergreen.yml [hunk 5]
```diff
@@ -50,6 +54,31 @@ functions:
...
+      params:
+        file: expansion.yml
+    - command: shell.exec
+      params:
+        script: |
+          ${PREPARE_SHELL}
+
+          # Fetch NVM and install it into this task's .nvm directory
+          # Once downloaded, source nvm and install yarn
+          git clone https://github.com/nvm-sh/nvm.git "$NVM_DIR"
```
Comment (khelif96) [inline .evergreen.yml:73]: Might be worthwhile asking mathew to see if we can prioritize the build NVM ticket. And also open a follow-up ticket that's linked to it to use it.

File: .evergreen.yml [hunk 6]
```diff
@@ -230,63 +245,45 @@ functions:
...
+        aws_key: ${aws_key}
+        aws_secret: ${aws_secret}
+        local_files_include_filter:
+          ["spruce/storybook-static/**/*.js.map"]
+        remote_file: spruce/${task_id}/storybook/
+        bucket: mciuploads
+        content_type: application/json
+        permissions: public-read
+        display_name: storybook_assets_map_
+        preserve_path: true
```
Comment (khelif96) [inline .evergreen.yml:286]: Great job using `preserve_path` flag we can probably remove this condition in the storybook config.
https://github.com/evergreen-ci/spruce/blob/5e564b68cbda0ea490c52d15425d702b50b74fc0/.storybook/main.js#L47-L49

**Review by khelif96 — APPROVED** [review]
Great job

File: .evergreen.yml [hunk 7]
```diff
@@ -230,63 +245,45 @@ functions:
...
+      params:
+        aws_key: ${aws_key}
+        aws_secret: ${aws_secret}
+        local_files_include_filter:
+          ["spruce/storybook-static/**/*.js", "spruce/storybook-static/**/*.mjs"]
+        remote_file: spruce/${task_id}/storybook/
+        bucket: mciuploads
+        content_type: application/javascript
+        permissions: public-read
+        display_name: storybook_assets_js_
```
Comment (minnakt) [inline .evergreen.yml:272]: I have a huge pet peeve over the blank space inserted between the display name and the file name  🫠 I would not be opposed if you wanted to change or remove the display names (optional‼️)

Screen Shot 2023-02-06 at 4 43 19 PM

**Review by minnakt — APPROVED** [review]
(mario voice) wahoo!

File: scripts/prepare-shell.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+export PROJECT_DIRECTORY="$(pwd)"
```
Comment (rtimmons) [inline scripts/prepare-shell.sh:3]: this requires the caller to know the right cwd to call the script. Consider doing something similar to what genny does:

https://github.com/mongodb/genny/blob/58540d3fc8ed01c16364dfaf254225226c5a364c/run-genny#L10-L13

File: scripts/prepare-shell.sh [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+cat <<EOT > expansion.yml
+PREPARE_SHELL: |
+    export PROJECT_DIRECTORY="$PROJECT_DIRECTORY"
+    export NVM_DIR="$NVM_DIR"
+
+    if [ -d "${NVM_DIR}" ]; then
+      cd $PROJECT_DIRECTORY
+      [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+      cd -
```
Comment (rtimmons) [inline scripts/prepare-shell.sh:14]: consider `pushd` and `popd` such that if you need to expand this script and `cd` to other locations in the future the `cd -` won't break.

File: scripts/prepare-shell.sh [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+export PROJECT_DIRECTORY="$(pwd)"
+export NVM_DIR="$PROJECT_DIRECTORY/.nvm"
+
+cat <<EOT > expansion.yml
+PREPARE_SHELL: |
+    export PROJECT_DIRECTORY="$PROJECT_DIRECTORY"
+    export NVM_DIR="$NVM_DIR"
+
+    if [ -d "${NVM_DIR}" ]; then
+      cd $PROJECT_DIRECTORY
```
Comment (rtimmons) [inline scripts/prepare-shell.sh:12]: this needs to be quoted
Please ensure this script passes `shellcheck`.

**Review by rtimmons — COMMENTED** [review]
Just a drive-by. Not blocking.

File: .evergreen.yml (hunk 6 above)
Comment (sophstad) [reply .evergreen.yml:283]: Interestingly removing this breaks the Storybook iframe's file paths!

File: cypress/integration/hosts/hosts_sorting.ts [hunk 11]
```diff
@@ -15,7 +15,8 @@ const sortByTests = [
       "build10.ny.cbi.10gen.cc",
     ],
   },
-  {
+  /* TODO EVG-18907: Re-enable distro sort upon completion of EVG-18900
```
Comment (khelif96) [inline cypress/integration/hosts/hosts_sorting.ts:18]: https://github.com/evergreen-ci/evergreen/pull/6255#pullrequestreview-1293333020

@ybrill  fixed this you should be able to re enable the tests @sophstad

**Pull request merged**